server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java

```diff
@@ -162,7 +170,12 @@ public TItem apply(ShardedRequests<TReq, TItem> shardedRequests, Row row, Boolea
 
             RowSourceInfo rowSourceInfo = RowSourceInfo.emptyMarkerOrNewInstance(sourceUri, lineNumber);
             ShardLocation shardLocation = getShardLocation(indexName, id, routing);
+            assert item instanceof ShardUpsertRequest.Item : "IndexItem must be instance of ShardUpsertRequest.Item";
+            IndexItem indexItem = (IndexItem) item;
             if (shardLocation == null) {
+                // Validation is done before creating an index in order to ensure
+                // that no "bad partitions" will be left behind in case of validation failure.
+                constraintsChecker.accept(indexNameResolver.get(), indexItem);
```

Comment:
Even that it currently on used by this, I don't see a logical reason to require it to be of type `IndexItem`.
```suggestion
            if (shardLocation == null) {
                // Validation is done before creating an index in order to ensure
                // that no "bad partitions" will be left behind in case of validation failure.
                if (item instanceof IndexItem indexItem) {
                    constraintsChecker.accept(indexNameResolver.get(), indexItem);
                }
```